File: backend/src/main/java/com/gxs/bank/controller/CardController.java
package com.gxs.bank.controller;

import com.gxs.bank.dto.request.CardApplyRequest;
import com.gxs.bank.dto.response.ApiResponse;
import com.gxs.bank.model.Card;
import com.gxs.bank.model.User;
import com.gxs.bank.service.CardService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.UUID;

@RestController
@RequestMapping("/api/cards")
@RequiredArgsConstructor
public class CardController {

    private final CardService cardService;

    @PostMapping("/apply")
    public ResponseEntity<ApiResponse<Card>> apply(
            @AuthenticationPrincipal User user,
            @Valid @RequestBody CardApplyRequest request) {
        Card card = cardService.applyForCard(user.getId(), request);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.ok("Card application submitted for approval", card));
    }

    @GetMapping
    public ResponseEntity<ApiResponse<List<Card>>> getCards(@AuthenticationPrincipal User user) {
        List<Card> cards = cardService.getUserCards(user.getId());
        return ResponseEntity.ok(ApiResponse.ok(cards));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<Card>> getCard(
            @PathVariable UUID id, @AuthenticationPrincipal User user) {
        Card card = cardService.getCard(id, user.getId());
        return ResponseEntity.ok(ApiResponse.ok(card));
    }

    @PutMapping("/{id}/freeze")
    public ResponseEntity<ApiResponse<Card>> toggleFreeze(
            @PathVariable UUID id, @AuthenticationPrincipal User user) {
        Card card = cardService.toggleFreeze(id, user.getId());
        return ResponseEntity.ok(ApiResponse.ok("Card status updated", card));
    }

    @PutMapping("/{id}/settings")
    public ResponseEntity<ApiResponse<Card>> updateSettings(
            @PathVariable UUID id,
            @AuthenticationPrincipal User user,
            @RequestBody Map<String, Object> settings) {
        Card card = cardService.updateSettings(id, user.getId(), settings);
        return ResponseEntity.ok(ApiResponse.ok("Card settings updated", card));
    }
}

File: backend-python/com/gxs/bank/controller/CardController.py
from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse

from com.gxs.bank.config.database import get_db
from com.gxs.bank.config.serialization import serialize
from com.gxs.bank.dto.request.CardApplyRequest import CardApplyRequest
from com.gxs.bank.dto.response.ApiResponse import ok
from com.gxs.bank.security.JwtAuthenticationFilter import get_current_user
from com.gxs.bank.service.CardService import CardService


router = APIRouter(prefix="/api/cards", tags=["Card"])


@router.post("/apply")
def applyForCard(request: CardApplyRequest, current_user=Depends(get_current_user), db=Depends(get_db)):
    card = CardService(db).applyForCard(current_user.id, request)
    return JSONResponse(
        status_code=201,
        content=ok(serialize(card), "Card application submitted for approval"),
    )


@router.get("")
def getCards(current_user=Depends(get_current_user), db=Depends(get_db)):
    cards = CardService(db).getUserCards(current_user.id)
    return ok(serialize(cards))


@router.get("/{card_id}")
def getCard(card_id: str, current_user=Depends(get_current_user), db=Depends(get_db)):
    card = CardService(db).getCard(card_id, current_user.id)
    return ok(serialize(card))


@router.put("/{card_id}/freeze")
def toggleFreeze(card_id: str, current_user=Depends(get_current_user), db=Depends(get_db)):
    card = CardService(db).toggleFreeze(card_id, current_user.id)
    return ok(serialize(card), "Card status updated")


@router.put("/{card_id}/settings")
def updateSettings(card_id: str, settings: dict, current_user=Depends(get_current_user), db=Depends(get_db)):
    card = CardService(db).updateSettings(card_id, current_user.id, settings)
    return ok(serialize(card), "Card settings updated")
